rust/actyx/ax-aql/src/language/types.rs

```diff
@@ -8,17 +8,258 @@ use crate::{
+                        }
+
+                        Ok(Type::Atom(TypeAtom::Record(
+                            fields
+                                .into_iter()
+                                .map(|(label, ty)| (label.clone(), ty.clone()))
+                                .collect::<Vec<(Label, Type)>>()
+                                .try_into()
+                                .unwrap(),
+                        ))
+                        .collapse()?)
+                    }
+                    _ => {
+                        if a.is_supertype_of(&b) {
+                            return Ok(b);
+                        }
+
+                        Err(CollapseError(vec![format!("{:?} and {:?} is of different type", a, b)]))
+                    }
+                }
+            }
+            Type::Union(x) => {
+                let flattened_collapsed = Type::flatten_union_collapsing(x.as_ref())?;
+                Ok(Type::union(flattened_collapsed.into_iter()))
+            }
+            Type::Atom(TypeAtom::Record(fields)) => {
+                let mut errors = vec![];
+                let fields = fields
+                    .into_iter()
+                    .cloned()
+                    .filter_map(|(label, ty)| match ty.collapse() {
+                        Ok(ty) => Some((label, ty)),
+                        Err(err) => {
+                            errors.push(err);
+                            None
+                        }
+                    })
+                    .collect::<BTreeMap<_, _>>();
+
+                if !errors.is_empty() {
+                    Err(CollapseError::join(errors))
+                } else {
+                    Ok(Type::Atom(TypeAtom::Record(
+                        fields.into_iter().collect::<Vec<_>>().try_into().unwrap(),
+                    )))
+                }
+            }
+            _ => Ok(self),
+        }
+    }
+
+    /// Calculate supertype without collapsing
+    fn is_supertype_of(&self, other: &Type) -> bool {
+        match (self, other) {
+            (Type::Atom(a), Type::Atom(b)) => match (a, b) {
+                (TypeAtom::Bool(None), TypeAtom::Bool(Some(_))) => true,
+                (TypeAtom::Number(None), TypeAtom::Number(Some(_))) => true,
+                (TypeAtom::String(None), TypeAtom::String(Some(_))) => true,
+                (TypeAtom::Record(a), TypeAtom::Record(b)) => {
+                    let a = a.iter().collect::<BTreeSet<_>>();
+                    let b = b.iter().collect::<BTreeSet<_>>();
+
+                    a.intersection(&b).cloned().collect::<BTreeSet<_>>() == b
+                }
+                (TypeAtom::Universal, _) => true,
+                _ => false,
+            },
+            (Type::Union(a), Type::Union(b)) => {
+                let a = Type::flatten_union(a.as_ref());
+                let b = Type::flatten_union(b.as_ref());
+
+                a.intersection(&b).cloned().collect::<BTreeSet<_>>() == b
```

Comment:
I'm not convinced about this, here's a counter example:


* Let `a` be `NUMBER | BOOLEAN`
* Let `b` be `NUMBER(10) | NULL`
* `a & b => NUMBER(10)`
* `(a & b) != b`

Let me know if I'm missing something